share/static/js/chat.js

```diff
@@ -1,65 +1,123 @@
-  var username_list = [];
   $(document).ready(function(){
-    socket = io.connect('http://' + document.domain + ':' + location.port + '/chat');
+    var socket = io.connect('http://' + document.domain + ':' + location.port + '/chat');
+
+    // Store current username received from app context
     var current_username = $('#username').val();
+
+    // On browser connect, emit a socket event to be added to 
+    // room and assigned random username
     socket.on('connect', function() {
       socket.emit('joined', {});
     });
+
+    // Triggered on any status change by any user, such as some
+    // user joined, or changed username, or left, etc.
     socket.on('status', function(data) {
-      $('#chat').append('<p><small><i>' + sanitizeHTML(data.msg) + '</i></small></p>');
-      if (data.user && current_username !== data.user) {
-        $('#user-list').append('<li>' + sanitizeHTML(data.user) + '</li>')
-        username_list.push(data.user);
-      }
-      if (data.new_name && current_username !== data.new_name) {
-        last_username = current_username;
-        current_username = data.new_name;
-        username_list[username_list.indexOf(last_username)] = current_username;
-        $('#user-list li').each(function(key, value) {
-          if ($(value).text() === data.old_name) {
-            $(value).html(sanitizeHTML(current_username));
-          }
-        })
-      }
-      $('#chat').scrollTop($('#chat')[0].scrollHeight);
-    });
-    socket.on('update_list', function(data) {
-      if (username_list.indexOf(data.name) === -1 &&
-        current_username !== data.name &&
-        last_username !== data.name
-      ) {
-        username_list.push(data.name);
-        $('#user-list').append('<li>' + sanitizeHTML(data.name) + '</li>')
-      }
-      $('#chat').scrollTop($('#chat')[0].scrollHeight);
+      addMessageToRoom(data, current_username, 'status');
     });
+
+    // Triggered when message is received from a user. Even when sent
+    // by self, it get triggered after the server sends back the emit.
     socket.on('message', function(data) {
-      $('#chat').append('<p>' + sanitizeHTML(data.msg) + '</p>');
-      $('#chat').scrollTop($('#chat')[0].scrollHeight);
+      addMessageToRoom(data, current_username, 'chat');
     });
+
+    // Trigger new message on enter or click of send message button.
     $('#new-message').on('keypress', function(e) {
       var code = e.keyCode || e.which;
       if (code == 13) {
         emitMessage(socket);
       }
     });
-    $('#send-button').on('click', emitMessage);
+    $('#send-button').on('click', function(e) {
+      emitMessage(socket);
+    });
+
+    // Update username
     $('#update-username').on('click', function() {
       var username = $('#username').val();
+      current_username = username;
       socket.emit('update_username', {username: username});
     });
+
+    // Show warning of losing data
+    $(window).on('beforeunload', function (e) {
```

Comment:
I am not 100% sure this is a good idea. Bit of conflicted here. On one hand, I feel it's important to give users a warning that closing or refreshing will lead them to lose all messages. On the other hand, if someone refreshed too many times (don't know why someone would do that), or every time someone closes the tab, this warning will be shown. So that might get frustrating. @holantonela @emmapeel2 would really love some feedback from you on this.